---
title: Genome Repeats Identification
layout: single
header:
  overlay_color: "444444"
  overlay_image: /assets/images/dna.jpg
---

# Repeats Identification

Repeats in genome sequence  can confuse gene annotation, and also hinder the process of genome assembly. It is crucial to annotate and classify them correctly in genome sequences.
the proportion of repeats in the genome can differ widely, ranging from a few percent (3% in the yeast Saccharomyces cerevisiae) to a huge proportion encompassing almost the entire genome (>80% in the maize). The human genome itself is particularly rich in repeats which make up about 45% of the genome.


## Index

  * [Helitron Identification in a Genome Sequence](Helitron_Scanner.md)
  * [DNA Transposon Annotation with Inverted-Repeats Finder](InvertedRepeatsFinderForDNATransposonAnnotation.md)
  * [LTR Retrotransposon Annotation with LTR-Finder](LTRFinder.md)  
  * [Repeat Annotation from Next-gen Sequencing Reads Using RepeatExplorer](RepeatExplorer.md)
  * [De-Novo Repeat Identification and Annotation from Genome Assemblies using RepeatModeler and RepeatMasker](RepeatModeler_RepeatMasker.md)
  * [Tandem Duplication Annotation in a Genome Assembly Using Mummer and RedTandem](Tandem_Duplication_Detection.md)
  * [How to assess structural variation in your genome, and identify jumping transposons](How_To_Find_Transposon_Jumps_in_Reads.md)
